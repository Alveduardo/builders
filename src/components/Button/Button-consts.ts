export const BUTTON = {
  TEST_ID: {},
};
