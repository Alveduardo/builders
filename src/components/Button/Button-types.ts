import { TouchableHighlightProps, ViewStyle } from 'react-native';
import { Color } from '../../utils/colors/colors-types';

export type ButtonSize = 'regular';
export type ButtonKind = 'primary' | 'secondary';

interface DefaultStyles {
  containerStyle: ViewStyle;
}

export type SizeStyle = {
  [key in ButtonSize]: {
    buttonStyle: {
      height: number;
    };
  };
};

export type KindStyle = {
  [key in ButtonKind]: {
    buttonStyle: ViewStyle;
    textStyle: {
      color: string;
    };
  };
};

export type StateStyle = {
  onPress: {
    [key in ButtonKind]: {
      buttonStyle: ViewStyle;
      textStyle: {
        color: string;
      };
    };
  };
  disabled: {
    [key in ButtonKind]: {
      buttonStyle: ViewStyle;
      textStyle: {
        color: string;
      };
    };
  };
};

export interface ButtonStyles {
  defaultStyles: DefaultStyles;
  sizeStyles: SizeStyle;
  kindStyles: KindStyle;
  stateStyles: StateStyle;
}

interface StateConfig {
  underlayColor: Color;
}

interface MethodsConfig {
  onPressIn: () => void;
  onPressOut: () => void;
}

interface StylesConfig {
  buttonStyle: ViewStyle;
  textColor: Color;
}

export interface ButtonConfig {
  state: StateConfig;
  methods: MethodsConfig;
  styles: StylesConfig;
}

export interface ButtonProps extends TouchableHighlightProps {
  /**
   * The label inside the button.
   */
  label: string;

  /**
   * The function that is invoked when the button is pressed.
   */
  onPress?: () => void;

  /**
   * Possible button size values.
   */
  size?: ButtonSize;

  /**
   * Possible button color values.
   */
  kind?: ButtonKind;

  /**
   * The disabled state.
   */
  disabled?: boolean;

  /**
   * The button's width. The possible values are `'auto'`, `'100%'`, or a numeric value representing the amount of pixels.
   */
  width?: number | string;
}
