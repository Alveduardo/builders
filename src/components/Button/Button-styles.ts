import { COLORS } from '../../utils/colors/colors-consts';
import { ButtonStyles } from './Button-types';

export const buttonStyles: ButtonStyles = {
  defaultStyles: {
    containerStyle: {
      width: '100%',
      borderRadius: 999,
      alignItems: 'center',
      justifyContent: 'center',
    },
  },
  sizeStyles: {
    regular: {
      buttonStyle: {
        height: 50,
      },
    },
  },
  kindStyles: {
    primary: {
      buttonStyle: {
        backgroundColor: COLORS.SECONDARY,
      },
      textStyle: {
        color: '#FFF',
      },
    },
    secondary: {
      buttonStyle: {
        backgroundColor: 'transparent',
        borderColor: COLORS.SECONDARY,
        borderWidth: 3,
      },
      textStyle: {
        color: COLORS.SECONDARY,
      },
    },
  },
  stateStyles: {
    onPress: {
      primary: {
        buttonStyle: {
          backgroundColor: COLORS.PRIMARY,
        },
        textStyle: {
          color: '#FFF',
        },
      },
      secondary: {
        buttonStyle: {
          backgroundColor: COLORS.SECONDARY,
          borderColor: COLORS.SECONDARY,
          borderWidth: 3,
        },
        textStyle: {
          color: '#FFF',
        },
      },
    },
    disabled: {
      primary: {
        buttonStyle: {
          backgroundColor: 'gray',
        },
        textStyle: {
          color: '#FFF',
        },
      },
      secondary: {
        buttonStyle: {
          backgroundColor: 'transparent',
          borderColor: 'gray',
          borderWidth: 3,
        },
        textStyle: {
          color: 'gray',
        },
      },
    },
  },
};
