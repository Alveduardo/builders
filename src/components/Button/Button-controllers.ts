import { useState } from 'react';
import { TextStyle, ViewStyle } from 'react-native';

import { ButtonConfig, ButtonProps } from './Button-types';
import { buttonStyles } from './Button-styles';

import { Color } from '../../utils/colors/colors-types';
import { setupStyleProp } from '../../utils/style/style-utils';

export const useButtonConfig = ({ width, size, kind, disabled, style }: Partial<ButtonProps>): ButtonConfig => {
  const [isPressed, setIsPressed] = useState(false);

  const {
    defaultStyles: { containerStyle },
    sizeStyles: {
      [size!]: { buttonStyle: buttonSizeStyle },
    },
    kindStyles: {
      [kind!]: { buttonStyle: buttonKindStyle, textStyle: textKindStyle },
    },
    stateStyles: {
      onPress: {
        [kind!]: { buttonStyle: buttonPressedStyle, textStyle: textPressedStyle },
      },
      disabled: {
        [kind!]: { buttonStyle: buttonDisabledStyle, textStyle: textDisabledStyle },
      },
    },
  } = buttonStyles;

  let buttonStyle: ViewStyle = { ...containerStyle, ...buttonSizeStyle, ...buttonKindStyle },
    textStyle: TextStyle = { ...textKindStyle };

  if (isPressed) {
    buttonStyle = { ...buttonStyle, ...buttonPressedStyle };
    textStyle = { ...textStyle, ...textPressedStyle };
  }
  if (disabled) {
    buttonStyle = { ...buttonStyle, ...buttonDisabledStyle };
    textStyle = { ...textStyle, ...textDisabledStyle };
  }
  if (width) {
    buttonStyle.width = width;
  }

  const underlayColor = buttonPressedStyle.backgroundColor as Color;

  const onPressIn = () => setIsPressed(true);

  const onPressOut = () => setIsPressed(false);

  buttonStyle = setupStyleProp(style, buttonStyle) as ViewStyle;

  return {
    state: { underlayColor },
    methods: { onPressIn, onPressOut },
    styles: { buttonStyle, textColor: textStyle.color as Color },
  };
};
