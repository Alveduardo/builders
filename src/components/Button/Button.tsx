import React, { useState } from 'react';
import { TouchableHighlight } from 'react-native';

import { ButtonProps } from './Button-types';
import { useButtonConfig } from './Button-controllers';

import Label from '../Label';
import { BUTTON } from './Button-consts';

const Button = ({
  label,
  width = '100%',
  size = 'regular',
  kind = 'primary',
  disabled = false,
  onPress = () => {},
  style = undefined,
  ...rest
}: ButtonProps): JSX.Element => {
  const {
    state: { underlayColor },
    methods: { onPressIn, onPressOut },
    styles: { buttonStyle, textColor },
  } = useButtonConfig({
    width,
    size,
    kind,
    disabled,
    style,
  });

  return (
    <TouchableHighlight
      testID={BUTTON.TEST_ID.CONTAINER}
      {...rest}
      {...{
        style: buttonStyle,
        disabled,
        onPress,
        onPressIn,
        onPressOut,
        underlayColor,
      }}
    >
      <Label color={textColor}>{label}</Label>
    </TouchableHighlight>
  );
};

export default Button;
