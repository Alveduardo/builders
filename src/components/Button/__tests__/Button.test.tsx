import React from 'react';
import { render } from '@testing-library/react-native';
import Button from '../Button';

describe('Button ->', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error').mockImplementation(jest.fn());
  });
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  describe('Snapshot ->', () => {
    it('Deve corresponder ao componente com as props default', () => {
      const tree = render(<Button label='test' />).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('Deve corresponder ao componente com as props mais utilizadas', () => {
      const tree = render(<Button label='test' />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
