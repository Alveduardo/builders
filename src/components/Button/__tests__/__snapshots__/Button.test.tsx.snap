// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button -> Snapshot -> Deve corresponder ao componente com as props default 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#036f93",
      "borderRadius": 999,
      "height": 50,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    rest={Object {}}
    style={
      Array [
        Object {
          "color": "#FFF",
          "fontSize": 16,
        },
        undefined,
      ]
    }
    testID="LABEL_CONTAINER_TEST_ID"
  >
    test
  </Text>
</View>
`;

exports[`Button -> Snapshot -> Deve corresponder ao componente com as props mais utilizadas 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#036f93",
      "borderRadius": 999,
      "height": 50,
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <Text
    rest={Object {}}
    style={
      Array [
        Object {
          "color": "#FFF",
          "fontSize": 16,
        },
        undefined,
      ]
    }
    testID="LABEL_CONTAINER_TEST_ID"
  >
    test
  </Text>
</View>
`;
